new lint: `char_indices_as_byte_indices`

y21 opened the pull request:
Closes #10202.

This adds a new lint that checks for uses of the `.chars().enumerate()` position in a context where a byte index is required and suggests changing it to use `.char_indices()` instead.

I'm planning to extend this lint to also detect uses of the position in iterator chains, e.g. `s.chars().enumerate().for_each(|(i, _)| s.split_at(i));`, but that's for another time

----------------
 
changelog: new lint: `chars_enumerate_for_byte_indices`

y21 commented on clippy_lints/src/loops/chars_enumerate_for_byte_indices.rs:
```diff
@@ -0,0 +1,138 @@
+        && let Some(method_id) = cx.typeck_results().type_dependent_def_id(iterable.hir_id)
+        && cx.tcx.is_diagnostic_item(sym::enumerate_method, method_id)
+        && let ExprKind::MethodCall(chars_segment, recv, _, chars_span) = recv.kind
+        && cx.typeck_results().expr_ty_adjusted(recv).peel_refs().is_str()
+        && chars_segment.ident.name.as_str() == "chars"
```
Latest nightly has `str::chars` as a diagnostic item but our pinned nightly isn't there yet. Might be able to use that if the next sync happens soon enough

y21 commented on tests/ui/chars_enumerate_for_byte_indices.rs:
```diff
@@ -0,0 +1,59 @@
+        let _ = prim[..prim.floor_char_boundary(idx)];
+        //~^ chars_enumerate_for_byte_indices
+
+        // can't use #[expect] here because the .fixed file will still have the attribute and create an
+        // unfulfilled expectation, but make sure lint level attributes work on the use expression:
+        #[allow(clippy::chars_enumerate_for_byte_indices)]
+        let _ = prim[..idx];
```
this is a fun one, I wonder if that's something that could be fixed in uitest, like removing `#[expect]` attributes in the .fixed file 🤔

y21 commented on clippy_lints/src/loops/mod.rs:
````diff
@@ -714,6 +715,48 @@ declare_clippy_lint! {
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub CHARS_ENUMERATE_FOR_BYTE_INDICES,
+    correctness,
````
The pattern *is* technically fine if you know what your strings are (like the description mentions) so it's not always 'outright wrong' like the usual correctness lints, but the fix is also really simple so 🤷‍♂️

Centri3 replied:
As said on Zulip, we could say that `bytes` should always be used as a replacement if it expects ASCII. I don't think there are any downsides to that. I think blocking compilation on those cases is better than not doing so for where it could actually be a problem.

y21 reviewed (commented):
Some notes for reviewers

Boshen commented:
Thank you for working on this. 

The linter we are working on (oxlint) has encountered dozens of crashes because of this, and there were no ways to forbid such usages.

Centri3 reviewed (approved):
This all looks fine to me. I'll open the FCP in a moment

Centri3 commented:
Oh one thing I can think of, this kind of analysis would definitely ideally make sense as a MIR lint, it would probably be simpler in some ways, but this works and MIR is incredibly annoying. Not sure what the stance has been on those anyway atm.

bors commented:
:umbrella: The latest upstream changes (presumably 1f966e98db0c83fc76657fbc0bd50d11526787b1) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Centri3 commented on clippy_lints/src/loops/chars_enumerate_for_byte_indices.rs:
```diff
@@ -0,0 +1,138 @@
+            && let PatKind::Binding(_, binding_id, ..) = pat.kind
+        {
+            // Destructured iterator element `(idx, _)`, look for uses of the binding
+            for_each_expr(cx, body, |expr| {
```
Any reason `for_each_local_use_after_expr` can't be used?

y21 replied:
Which expression's HirId do we pass to it, given we want to visit the whole loop's body (only)? The loop expression or iterable expression is outside the block so can't be used

Centri3 replied:
Hmm yeah fair. Missed that and unfortunately that won't work

y21 commented:
Added a note for `str::bytes` and also renamed the lint to be slightly more general in case we have other sources of char indices in the future as Jarcho suggested on zulip. Left them as separate commits for now so it's easier to review but will squash once you think it's ok

Centri3 commented on clippy_lints/src/loops/char_indices_as_byte_indices.rs:
```diff
@@ -0,0 +1,138 @@
+            "passing a character position to a method that expects a byte index"
+        },
+        ExprKind::Index(target, ..)
+            if is_string_like(cx.typeck_results().expr_ty_adjusted(target).peel_refs())
```
nit: I think the reason this isn't used in both arms is a bit non-obvious at first, especially if the comment on `BYTE_INDEX_METHODS` isn't seen. It could perhaps just point to what's already written there. That's about all.

Centri3 commented on clippy_lints/src/loops/char_indices_as_byte_indices.rs:
```diff
@@ -0,0 +1,138 @@
+        ExprKind::MethodCall(segment, recv, ..)
+            if cx.typeck_results().expr_ty_adjusted(recv).peel_refs().is_str()
+                && BYTE_INDEX_METHODS.contains(&segment.ident.name.as_str())
+                && eq_expr_value(cx, chars_recv, recv) =>
```
I could find this check a bit suspicious in that I can't see anywhere it checks that the string hasn't been operated on or cloned (if there is then this can be ignored). In that case, the `&mut self` methods could work depending on how `eq_expr_value` works (as contrived as it is): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e27e7dac1f146be9a020f1c9fc822b7

y21 replied:
So I assume you mean this in the "false negative" sense in that your linked example could have a warning but doesn't. I suppose we could "skip" `.clone()` calls and just look at the cloned strings. Although I guess it's very unlikely for something like `string.clone().remove(idx);` in your example to exist in real code, without assigning the clone to some temporary variable first. Handling smth like
```rs
for (idx, _) in s.clone().chars().enumerate() {
  let s = s.clone();
  s.remove(idx);
}
```
could be doable I suppose, but I feel like that would complicate this lint a fair amount.


Alternatively, I do wonder if we actually need to check which string is being used at all and whether we could lint even if you use some other random/unknown string. I'm not sure if that could lead to false positives, but it would allow linting cases where you know as an invariant that two strings have the same length and then use the index into the other string or something, like
```rs
fn f(s1: &str, s2: &str) {
  assert_eq!(s1.len(), s2.len()); // invariant

  for (idx, _) in s1.chars().enumerate() {
    do_something(&s2[idx..]); // Could make sense to lint even if it isn't the same string?
  }
}
```

Centri3 replied:
> So I assume you mean this in the "false negative" sense in that your linked example could have a warning but doesn't.

I mean the opposite. iirc, when I tested it, this emitted a warning. I don't think it should. The issue I'm referencing is that if the two receivers are the same, no matter if it's just a local use or not, it'll lint. Thus you get this odd behavior.

This also means that the idea of the `String` methods being impossible to use in this way isn't necessarily correct in the comment, but it would be if this is changed

y21 replied:
The example you linked shouldn't have a warning (and doesn't, just tested it out) since `eq_expr_value()` denies any side effects, so `string.clone()` should nevetr be considered equal to any other expression as it contains a method call. I'm surprised that you saw it emit a warning here

Centri3 replied:
> since `eq_expr_value()` denies any side effects

Right, good to know!

tbh, I have no idea. I'm not sure if it was a false memory at the time or whether there was another case I was mistaking it with (but an erroneous one nonetheless). I doubt it was the latter. I still find this check, *wrong?*, as if I am missing some case, but if you think it works then I will trust your judgement

r=me after a rebase alongside w/e you think about the comment nit

y21 replied:
I added this as a test case to make sure it really doesn't lint there along with the comment (last commit). Does that help? I can squash all the commits once/if you think that's fine.

Centri3 replied:
Sounds good. Let's not delay this further :p

Centri3 reviewed (approved):
Looks good to me :)

Centri3 commented:
Hey @y21, can you take a look at this again? Is there anything from my review that doesn't make sense/you're stuck on?
